src/fidesops/task/refine_target_path.py

```diff
@@ -0,0 +1,147 @@
+import logging
+from typing import Dict, Any, List, Union
+
+from fidesops.graph.config import FieldPath
+
+Level = Union[str, int]
+DetailedPath = List[
+    Level
+]  # A more detailed path to a field, potentially containing indices
+
+logger = logging.getLogger(__name__)
+
+
+def build_incoming_refined_target_paths(
+    row: Dict[str, Any], query_paths: Dict[FieldPath, List[Any]]
+) -> List[DetailedPath]:
+    """
+    Return a list of more detailed path(s) to the matched data that caused that row to be returned. Runs
+    recursive `refine_target_path` for each FieldPath in query_paths. If there are no array paths, the paths
+    will not change.
+
+    :param row: Record retrieved from a dataset
+    :param query_paths: FieldPaths mapped to query values
+    :return: A list of lists containing more detailed paths to the matched data
+
+    :Example:
+    row = {
+        "A": [1, 2],
+        "B": 2,
+        "C": [{"D": 3, "E": 5}, {"D": 3, "E": 4}, {"D": 3, "E": 4}],
+        "G": 3,
+    }
+    incoming_paths= {FieldPath("A"): [2], FieldPath("C", "E"): [4], FieldPath("G"): [3]}
+
+    build_incoming_refined_target_paths(row, incoming_paths)
+    [["G"], ["A", 1], ["C", 1, "E"], ["C", 2, "E"]]
+    """
+    found_paths: List[DetailedPath] = []
+    for target_path, only in query_paths.items():
+        path = refine_target_path(row, list(target_path.levels), only)
+        if path:
+            if isinstance(path[0], list):
+                found_paths.extend(path)
+            else:
+                found_paths.append(path)
+    found_paths.sort(key=len)
+    return found_paths
+
+
+def refine_target_path(
+    row: Dict[str, Any], target_path: List[str], only: List[Any]
+) -> DetailedPath:  # Can also return a list of DetailedPaths if there are multiple matches.
+    """
+    Recursively modify the target_path to be more detailed path(s) to the referenced data. Instead of just strings,
+    the path will be expanded to include indices where applicable.
+
+    :param row: Record retrieved from a dataset
+    :param target_path: A list of strings representing the path to the desired field.
+    :param only: A list of values that were used to build the query.
+    :return: A list or a list of lists containing more detailed path(s) to the data in "only".  If there
+    was just one path, we return one list.
+
+    :Example:
+    In this example, path ["A", "B", "C"] points to two elements that match values "F" or "G".  We update
+    the path to insert the indices to locate the appropriate value.
+
+    refine_target_path({"A": {"B": [{"C": "D"}, {"C": "F"}, {"C": "G"}]}}, ["A", "B", "C"], only=["F", "G"])
+
+    [["A", "B", 1, "C"], ["A", "B", 2, "C"]]
+    """
```

Comment:
This method (which is a recursive method called as part of `filter_element_match`) should also be useful in making targeted edits for erasures when i get there.  I can take the field_path, build a more refined field path to specific element(s) in the record, mask just those specific values, and write the whole record back. In mongo, we can also write queries to update specific elements in arrays, but other datasets may not have this capability - in general I've added several Python methods to do some of this in code rather than at the database level.

Reply:
I use a lot of recursion in this PR to crawl through arrays and dictionaries of unknown depth - I generally try to avoid recursive solutions in Python - Python doesn't have tail call recursion, there's generally a way to implement it with an iterative solution, recursion is limited to 1000 function calls by default, and the language has a strong preference for clear, readable solutions.   

To help with the readable part, I've tried to add a lot of detail to the docstrings with examples.  There may be ways to simplify the recursion, but a lot of these pieces were added to cover different array use cases.